### DISCUSSION TITLE
Implement proper support of PEP 435 enums

### DISCUSSION THREAD

**Discussion by ethanfurman** [sqlalchemy/sqlalchemy#8934] [Ideas]
When `Enum` was introduced in Python 3.4 there was a *lot* of discussion.  One of the bigger discussions was about the relation of an enum name to its value.

The name/value issue was solved in two ways:

- enum members would have a `value` attribute that could hold whatever was necessary; and
- enums could subclass other data types, such as `int`, to more easily interoperate with libraries, functions, etc., that were already expecting a particular data type.

The expectation and intended use is that the enum member's name would be significant in Python code, and it's value, if needed at all, would be the interface to other systems -- such as databases.

Unfortunately, when `Enum` became supported in SqlAlchemy, that's not what happened: the member's name is what is persisted to the database, and its value is ignored.

This has led to several issues and workarounds:

- https://github.com/sqlalchemy/sqlalchemy/issues/3906
- https://github.com/sqlalchemy/sqlalchemy/issues/4180
- https://github.com/sqlalchemy/sqlalchemy/issues/6146

I would like to see PEP 435 enums supported directly and correctly without the extra effort currently required (such as callables to get the values to be persisted instead of the names), and I am happy to help with the effort.

**Comment by zzzeek** [comment]
As mentioned in #3906, the value of an enum can be any object.   Whereas SQLAlchemy's `Enum` datatype is oriented towards built-in database types such as PostgreSQL and MySQL ``ENUM``, which store only strings.

What is your proposal for an Enum datatype that can persist other kinds of values?  What SQL datatype would be used?

**Reply by ethanfurman** [reply]
My proposal is not about persisting other types of values, it's about users creating enums with a proper value (`str`, in the case of PostgreSQL and MySQL), and SqlAlchemy using those values.

For example, the docs have:

    class MyEnum(enum.Enum):
        one = "one"
        two = "two"
        three = "three"

Following the suggestions in the [enum docs](https://docs.python.org/3/howto/enum.html#enum-basic-tutorial), that would be:

    class MyEnum(enum.Enum):
        ONE = "one"
        TWO = "two"
        THREE = "three"

In Python 3.11, one could use the new `StrEnum` (whose `auto()` defaults to lowercase member names):

    class MyEnum(enum.StrEnum):
        ONE = auto()
        TWO = auto()
        THREE = auto()

Then anyone familiar with the stdlib `Enum` would expect, and get, the values of `'one'`, `'two'`, and `'three'` persisted in the database.

**Reply by zzzeek** [reply]
oh, I was all set to have a whole new hierarchy with `IntValueEnum`, `StringValueEnum`, etc.

as @CaselIT mentions we already support this, if you want a simple convenience parameter for "values_callable" hardcoded to string enums, that's not very controversial.   it can't be by default however one can set it up in the [type_annotation_map](https://docs.sqlalchemy.org/en/20/orm/declarative_tables.html#using-python-enum-types-in-the-type-map) with this setting, so that it works automatically.

**Comment by CaselIT** [comment]
Hi,

I don't think this is feasible to change the default behaviour of enum since this would have large impacts to user code. 
To implement this probably requires creating a new data type, or adding yet another flag to Enum which defaults to false. Considering the amount of effort that the current Enum requires in alembic, I personally don't have much appetite for yet another iteration of it.

That said, considering there is already a values_callable maybe a flag that sets a default values_callable could be something that's not too hard to implement and support
